easybuild/easyblocks/generic/cargo.py

```diff
@@ -339,38 +365,68 @@ def extract_step(self):
-            self.log.debug("Writting config.toml entry for vendored crates from crate.io")
-            write_file(config_toml, CONFIG_TOML_SOURCE_VENDOR.format(vendor_dir=vendor_dir), append=True)
-
-            # Tell cargo about the vendored git sources to avoid it failing with:
-            # Unable to update https://github.com/[...]
-            # can't checkout from 'https://github.com/[...]]': you are in the offline mode (--offline)
-            for (git_repo, rev), src in git_sources.items():
-                self.log.debug("Writting config.toml entry for git repo: %s rev %s", git_repo, rev)
-                src_dir = src['finalpath']
-                if os.path.dirname(src_dir) == vendor_dir:
-                    # Non-workspace sources are in vendor_dir
+            self._setup_offline_config(git_sources)
+
+    def _setup_offline_config(self, git_sources):
+        """Setup the configuration required for offline builds
+
+        :param git_sources: dict mapping (git_repo, rev) to extracted source
+                            rev is a tuple of (revision, branch) or just the revision for older easyconfigs
+        """
+        self.log.info("Setting up vendored crates for offline operation")
+        self.log.debug("Writting config.toml entry for vendored crates from crate.io")
+        config_toml = os.path.join(self.cargo_home, 'config.toml')
+        # Replace crates-io with vendored sources using build dir wide toml file in CARGO_HOME
+        write_file(config_toml, CONFIG_TOML_SOURCE_VENDOR.format(vendor_dir=self.vendor_dir))
+
+        # Tell cargo about the vendored git sources to avoid it failing with:
+        # Unable to update https://github.com/[...]
+        # can't checkout from 'https://github.com/[...]]': you are in the offline mode (--offline)
+
+        # "[patch.<repo>]" entries for git sources in the vendor folder
+        config_toml_git_patches = defaultdict(str)
+        # Unique revisions per repo
+        repo_revs = defaultdict(set)
+        for src, rev, _ in git_sources:
+            repo_revs[src].add(rev)
+        for (git_repo, rev, branch), src in git_sources.items():
+            src_dir = src['finalpath']
+            if os.path.dirname(src_dir) == self.vendor_dir:
+                # Non-workspace sources are in vendor_dir
+                # We want to specify the repo using the revision and optional branch used which is most reliable.
+                if branch is not self.BRANCH_MISSING or len(repo_revs[git_repo]) > 1:
+                    self.log.debug("Writing config.toml entry for git repo: %s branch %s, rev %s",
+                                   git_repo, branch, rev)
+                    branch_info = 'branch = "%s"\n' % branch if branch else '\n'
                     write_file(config_toml,
-                               CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev),
+                               CONFIG_TOML_SOURCE_GIT.format(url=git_repo, rev=rev) + branch_info,
```

Comment:
I prefer to have a different template for the repos with branch information. This currently splits the format of that configuration in two distant parts of the code (template in line 67, branch spec in line 407), which is not clear.

```suggestion
                               CONFIG_TOML_SOURCE_GIT_BRANCH.format(url=git_repo, rev=rev, branch=branch),
```

Reply:
Agreed. My thinking here was: We have a template with url and rev keys and if we have another one it would have url, rev, branch keys. So now the knowledge about the existing keys is non-local which isn't much better.

So with your suggestion we'd have `CONFIG_TOML_SOURCE_GIT` and `CONFIG_TOML_SOURCE_GIT_BRANCH` and decide on `if branch` which one to use. So something like:

```
template = CONFIG_TOML_SOURCE_GIT_BRANCH if branch else CONFIG_TOML_SOURCE_GIT
write_file(config_toml, template.format(url=git_repo, rev=rev, branch=branch))
```
`format` doesn't care about unused args, so that should be the cleanest.